### PR TITLE
Allow structure reuse if a declaration file within a package is updated

### LINKED ISSUE #47471: No structure reuse when a declaration file within a package is updated
# Bug Report

### 🔎 Search Terms

structure reuse, performance

### 🕗 Version & Regression Information

Current `main` (b53073bf66571b4c99f0a93b8f76fb06b36d8991)

- This is the behavior in every version I tried

### ⏯ Playground Link

Not applicable.

### 💻 Code

```ts
    it("successful if change affects a single module of a package", () => {
            const files = [
                { name: "/a.ts", text: SourceText.New("", "import {b} from 'b'", "var a = b;") },
                { name: "/node_modules/b/index.d.ts", text: SourceText.New("", "export * from './internal';", "") },
                { name: "/node_modules/b/internal.d.ts", text: SourceText.New("", "", "export const b = 1;") },
                { name: "/node_modules/b/package.json", text: SourceText.New("", "", JSON.stringify({ name: "b", version: "1.2.3" })) },
            ];

            const options: CompilerOptions = { target, moduleResolution: ModuleResolutionKind.NodeJs };
            const program1 = newProgram(files, ["/a.ts"], options);
            const program2 = updateProgram(program1, ["/a.ts"], options, files => {
                files[2].text = files[2].text.updateProgram("export const b = 2;");
            });
            assert.equal(program2.structureIsReused, StructureIsReused.Completely);
            const program1Diagnostics = program1.getSemanticDiagnostics(program1.getSourceFile("a.ts"));
            const program2Diagnostics = program2.getSemanticDiagnostics(program1.getSourceFile("a.ts"));
            assert.equal(program1Diagnostics.length, program2Diagnostics.length);
        });
```

### 🙁 Actual behavior

TypeScript determines that `StructureIsReused.Not` is in order for the change to the declaration file `node_modules/b/internal.d.ts`, because it notices that there's two files `node_modules/b/index.d.ts` and ``node_modules/b/internal.d.ts` corresponding with the same package name. This currently prevents structure reuse to detect changes to redirect source files, but here the declaration files are not involved in redirects.

The structure reuse logic for redirects was initially implemented in #16274, in which `PackageId` only consisted of `name` and `version` and the structure reuse check was based on `name`. Then in #18185 the `PackageId` structure was extended with `subModuleName` but this extension was not reflected in the structure reuse logic.

Because the structure reuse detection is based on having multiple files for the same package, structural reuse is possible when updating the declaration of a package when there is only a single declaration file present.

### 🙂 Expected behavior

The structure should be reused, as an update to a declaration file within a package should not be interpreted to be related to a redirect.

### PR DESCRIPTION
**Pull request by JoostK**
Closes #47471

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@sheetalkamat this seems worthy of signaling your attention for

File: src/testRunner/unittests/reuseProgramStructure.ts [hunk 1]
```diff
@@ -249,6 +249,25 @@ namespace ts {
...
+            ];
+
+            const options: CompilerOptions = { target, moduleResolution: ModuleResolutionKind.NodeJs };
+            const program1 = newProgram(files, ["/a.ts"], options);
+            const program2 = updateProgram(program1, ["/a.ts"], options, files => {
+                files[2].text = files[2].text.updateProgram("export const b = 2;");
+            });
+            assert.equal(program2.structureIsReused, StructureIsReused.Completely);
+            const program1Diagnostics = program1.getSemanticDiagnostics(program1.getSourceFile("a.ts"));
+            const program2Diagnostics = program2.getSemanticDiagnostics(program1.getSourceFile("a.ts"));
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/reuseProgramStructure.ts:267]:
```suggestion
            const program2Diagnostics = program2.getSemanticDiagnostics(program2.getSourceFile("a.ts"));
```

?

Comment (andrewbranch) [reply src/testRunner/unittests/reuseProgramStructure.ts:267]: Should be `===` but yeah 😅

Comment (JoostK) [reply src/testRunner/unittests/reuseProgramStructure.ts:267]: Good point I'll fix that tomorrow 👍

Comment (JoostK) [reply src/testRunner/unittests/reuseProgramStructure.ts:267]: Oh btw this is the case for all these tests so I'll fix those as well.

**Comment by andrewbranch** [comment]
This seems fine to me but I'll wait a bit to see if Sheetal wants to take a quick look

**Comment by sheetalkamat** [comment]
I can take a look next week and get back

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -192,6 +192,10 @@ namespace ts {
         return a === b || !!a && !!b && a.name === b.name && a.subModuleName === b.subModuleName && a.version === b.version;
     }
 
+    export function packageIdToPackageName({ name, subModuleName }: PackageId): string {
+        return subModuleName ? `${name}/${subModuleName}` : name;
+    }
+
     export function packageIdToString({ name, subModuleName, version }: PackageId): string {
         const fullName = subModuleName ? `${name}/${subModuleName}` : name;
         return `${fullName}@${version}`;
```
Comment (sheetalkamat) [inline src/compiler/utilities.ts:201]:
```suggestion
    export function packageIdToString(packageId: PackageId): string {
        return `${packageIdToPackageName(packageId)}@${packageId.version}`;
```

**Review by sheetalkamat — APPROVED** [review]
Change looks correct

**Comment by JoostK** [comment]
This is good to go from my side.

**Comment by JoostK** [comment]
Can this also be cherry-picked into `release-4.6` (i.e. the 4.6 beta), assuming that `main` is targeting 4.7 development?

**Comment by DanielRosenwasser** [comment]
The `main` branch actually does still target 4.6 until we hit 4.6 RC, so this change should make its way into TS 4.6 (or nightly if you need it sooner 🙂).

**Comment by JoostK** [comment]
Ah, excellent, that works for me :-) Thanks for the quick turnaround times on this PR, much appreciated!

**Pull request merged**